hal/stm32f373/i2c.c

```diff
@@ -1,391 +1,401 @@
 {
-    // sanity checks
-    int delay=100;
-    if (len < 1)
-        ///@todo invalid input parameters
-        return;
-    I2C_Timeout = I2C_LONG_TIMEOUT;
-    while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_BUSY) != RESET)
-    {
-        if((I2C_Timeout--) == 0) return -1;
-    }
-    sys_enter_critical_section();   // lock while changing things so an isr does not find a half setup write
-    if (i2c->write_buf != NULL || i2c->write_count != 0)
-        ///@todo write in progress already
-        goto done; 
-    // load the write info
-    i2c->write_buf = buf;
-    i2c->write_buf_len = len;
-    i2c->write_count = 0;
-    i2c->write_complete_cb = cb;
-    i2c->write_complete_param = param;
-    /*send slave address*/
-    I2C_TransferHandling(i2c->channel, device_address, len, I2C_AutoEnd_Mode, I2C_Generate_Start_Write);
-    I2C_Timeout = I2C_LONG_TIMEOUT;
-    while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_TXIS) == RESET)
-    {
-        if((I2C_Timeout--) == 0) return -1;
-    }
-    if (i2c->tx_dma)
-    {
-        // dma transfer
-        i2c->tx_dma_req.complete = i2c_tx_dma_complete;
-        i2c->tx_dma_req.complete_param = i2c;
-        i2c->tx_dma_req.dma = i2c->tx_dma;
-        i2c_dma_cfg(i2c, I2C_DMA_DIR_TX, &i2c->tx_dma_req, i2c->write_buf, i2c->write_buf_len);
-        I2C_DMACmd(i2c->channel, I2C_DMAReq_Tx, ENABLE);
-        dma_request(&i2c->tx_dma_req);
-    }    
-    else
-    {
-        // interrupt mode 
-        I2C_ITConfig(i2c->channel,I2C_IT_TXI, ENABLE);
-    }
+	// sanity checks
+	if (len < 1)
+		///@todo invalid input parameters
+		return -1;
+
+	sys_enter_critical_section();   // lock while changing things so an isr does not find a half setup write
+
+	uint32_t timeout = I2C_LONG_TIMEOUT;
+	while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_BUSY) != RESET)
+	{
+		if((timeout--) == 0)
+			goto done;
+	}
+
+	if (i2c->write_buf != NULL || i2c->write_count != 0)
+		///@todo write in progress already
+		goto done; 
+	// load the write info
+	i2c->write_buf = buf;
+	i2c->write_buf_len = len;
+	i2c->write_count = 0;
+	i2c->write_complete_cb = cb;
+	i2c->write_complete_param = param;
+	/*send slave address*/
+	I2C_TransferHandling(i2c->channel, device_address, len, I2C_AutoEnd_Mode, I2C_Generate_Start_Write);
+	timeout = I2C_LONG_TIMEOUT;
+
+	while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_TXIS) == RESET)
+	{
+		if((timeout--) == 0) 
+			goto done;
+	}
+
+	if (i2c->tx_dma)
+	{
+		// dma transfer
+		i2c->tx_dma_req.complete = i2c_tx_dma_complete;
+		i2c->tx_dma_req.complete_param = i2c;
+		i2c->tx_dma_req.dma = i2c->tx_dma;
+		i2c_dma_cfg(i2c, I2C_DMA_DIR_TX, &i2c->tx_dma_req, i2c->write_buf, i2c->write_buf_len);
+		I2C_DMACmd(i2c->channel, I2C_DMAReq_Tx, ENABLE);
+		dma_request(&i2c->tx_dma_req);
+	}    
+	else
+	{
+		// interrupt mode 
+		I2C_ITConfig(i2c->channel,I2C_IT_TXI, ENABLE);
+	}
 done:
-    sys_leave_critical_section();
-    return ;
+	sys_leave_critical_section();
+	return 0;
```

Comment:
again a note to future @othane (as Homer says, I don't envy that guy) ... but we should return the number of bytes placed into the write buffer or and error code (ie < 0)